docs/model-dev-guide/apis-howto/deepspeed/autotuning.rst

```diff
@@ -0,0 +1,303 @@
+parameters to be specified as command-line arguments, an additional helper function,
+:func:`~determined.pytorch.dsat.get_hf_args_with_overwrites`, is needed to create consistent
+HuggingFace arguments.
+
+Here is an example code snippet from a HuggingFace Trainer script that contains key pieces of
+relevant code:
+
+.. code:: python
+
+   from determined.integrations.huggingface import DetCallback
+   from determined.pytorch import dsat
+   from transformers import HfArgumentParser,Trainer, TrainingArguments,
+
+   hparams = self.context.get_hparams()
+   parser = HfArgumentParser(TrainingArguments)
+   args = sys.argv[1:]
+   args = dsat.get_hf_args_with_overwrites(args, hparams)
+   training_args = parser.parse_args_into_dataclasses(args, look_for_args_file=False)
+
+   det_callback = DetCallback(core_context, ...)
+   trainer = Trainer(args=training_args, ...)
+   with dsat.dsat_reporting_context(core_context, op=det_callback.current_op):
+       train_result = trainer.train(resume_from_checkpoint=checkpoint)
+
+.. important::
+
+   -  The ``dsat_reporting_context`` context manager shares the same initial
+      :class:`~determined.core.SearcherOperation` as the ``DetCallback`` instance through its
+      ``op=det_callback.current_op`` argument.
+
+   -  The entire ``train`` method of the HuggingFace trainer is now wrapped in the
+      ``dsat_reporting_context`` context manager.
+
+A full example which uses DeepSpeed Autotune with HuggingFace Trainer can be found in the
+``examples/hf_trainer_api`` `subdirectory within the Determined GitHub Repo
+<https://github.com/determined-ai/determined/tree/master/examples/hf_trainer_api>`__ .
+
+******************
+ Advanced Options
+******************
+
+The command-line entrypoint to ``dsat`` has various available options, some of them
+search-algorithm-specific. All available options for any given search method can be found as in
+
+.. code::
+
+   python3 -m determined.pytorch.dsat asha --help
+
+Below, we highlight particularly important flags and describe the search algorithms in some more
+detail.
+
+General Options
+===============
+
+The following options are available for every search method.
+
+-  ``--max-trials``: The maximum number of trials to run. Default: ``50``.
+
+-  ``--max-concurrent-trials``: The maximum number of trials that can run concurrently. Default:
+   ``16``.
+
+-  ``--max-slots``: The maximum number of slots that can be used concurrently. Defaults to ``None``,
+   i.e., there is no limit by default.
+
+-  ``--metric``: The metric to be optimized. Defaults to ``FLOPS-per-gpu``. Other available options
+   are ``throughput``, ``forward``, ``backward``, and ``latency``.
+
+-  ``--run-full-experiment``: If specified, after the ``dsat`` experiment has completed, a
+   ``single`` experiment will be launched using the specifications in the ``deepspeed.yaml``
+   overwritten with the best-found DS configuration parameters.
```

Comment:
A little tangential to this exact review, but I was thinking, do we not want to make this option a default? Shouldn't it be more like the other way around with it defaulting to running a full experiment and users could choose to not run this full experiment with something like `--dry-run` or something? I know it's a little out of scope but it reminded me

Reply:
Yeah, I think it's not an obvious call. I personally wouldn't really want to use that option very often because I would want to sanity check the result before launching the follow on run which may be very expensive.  I believe I followed the native DS AT behavior by not running the full experiment afterwards by default.